### ISSUE TITLE
Comments for Why is my function slow?

### ISSUE THREAD

**Issue by phinjensen** [EndPointCorp/end-point-blog#80]
Comments for https://www.endpoint.com/blog/2008/12/11/why-is-my-function-slow

To enter a comment:
1. Log in to GitHub
2. Leave a comment on this issue.

**Comment by phinjensen** [comment]
```
original author: Pavel Stěhule
date: 2008-12-11T16:27:00-05:00
```

your function from sample will be radically simplified in 8.4:

CREATE FUNCTION count_rentals_since_date_dynamic(date)
RETURNS BIGINT
LANGUAGE plpgsql
AS $body$
  BEGIN
    RETURN QUERY EXECUTE 'SELECT count(*) FROM rental WHERE rental_date > $1' USING $1;
  END;
$body$;

regards
Pavel Stehule

**Comment by phinjensen** [comment]
```
original author: Josh
date: 2008-12-11T16:48:00-05:00
```

It would be interesting (and lots of programming) to let a prepared statement cache a set of query plans, perhaps as many as one per filter step in the query, and choose between them at execution time based on the selectivity of each filter.

**Comment by phinjensen** [comment]
```
original author: Daniel Lyons
date: 2008-12-11T18:11:00-05:00
```

It's probably faster (in addition to being more portable and safer) to use SQL instead of PL/pgSQL, since you're not really doing anything but running an SQL query with it. In other words, something like:

CREATE FUNCTION count_rentals_since_date_dynamic(date)
RETURNS BIGINT
AS $body$
SELECT count(*) FROM rental WHERE rental_date > $1;
$body$ LANGUAGE sql;

**Comment by phinjensen** [comment]
```
original author: Greg Sabino Mullane
date: 2008-12-11T18:45:00-05:00
```

Pavel: Cool, 8.4 is going to be a nice release indeed. Thanks for the blog updates on what to look forward to!

Josh: Sounds tricky, I don't know if that would really gain much over just making a complete plan then and there.

Daniel: Yes, but this was just a simplified example, and most real-world functions with this problem (that I've come across) are plpgsql. I even thought about putting a SQL version of the function in the article, but decided to keep the article shorter and focused.

**Comment by phinjensen** [comment]
```
original author: Joshua Tolley
date: 2008-12-12T01:37:00-05:00
```

The multiple cached plans technique would definitely be tricky :) I'm expecting most of the parsing overhead would happen only once per query (rather than, say, once per cached plan), and that it would gracefully avoid catastrophic failures such as you demonstrate while adding little to execution time for a prepared query. But I can't back up that expectation, yet :)

**Comment by phinjensen** [comment]
```
original author: Chris
date: 2009-02-19T14:06:00-05:00
```

Hi, I'm a developer... not a DBA. We don't have a DBA, so I was tasked with figuring out what was going with some triggers in our Postgres 8.1 database. This article is the only thing I could find that helped. In fact, I started using strings/executions instead of just plain statements all over and cut execution time by over 90%. 

Anyway, it seems like caching of a bad plan for functions is a huge failure... and I'm a little shocked I didn't find any other references to it out there. Maybe I'm looking in the wrong place. Can anyone refer me to some documentation or something that points this postgres foible out? Or suggests this awesome solution as a best-practice?

Of course, the proof is in the pudding.. and the query is way better. But, still it'd be nice to read more about this.

I appreciate any input.

**Comment by phinjensen** [comment]
```
original author: Greg Sabino Mullane
date: 2009-02-24T14:02:00-05:00
```

Chris:

I don't think you'll find too much info out there. It's just a limitation of Postgres - it has to make a best one-time guess. For a lot of functions, this works out.

The official docs mention it as an almost aside, buried deep in here:

http://www.postgresql.org/docs/current/static/plpgsql-implementation.html#PLPGSQL-PLAN-CACHING

There is probably many posts on the Postgres mailing lists that have pointed out this trick as well.

**Comment by phinjensen** [comment]
```
original author: Robert
date: 2013-05-06T09:39:47-04:00
```

Thank you very much. That solved my problem!

**Comment by phinjensen** [comment]
```
original author: Anonymous
date: 2013-10-22T20:25:53-04:00
```

It sounds like a workaround that everybody uses and having to build a query dynamically just for this reason sounds silly. It makes the code not easy to read and adds complexity. I don't get why still was not implemented a better way to let the parser know that a new plan should be created for the query? Oracle has HINTS, there is no such thing here?

**Comment by phinjensen** [comment]
```
original author: Greg Sabino Mullane
date: 2013-11-04T22:16:38-05:00
```

Anon: There is some work lately on solving this problem, but without hints: more of an automatic reparsing as necessary. Hints are a hot topic but generally frowned upon in the Postgres world as being more trouble than they are worth.

**Comment by phinjensen** [comment]
```
original author: Greg Sabino Mullane
date: 2014-04-28T14:05:22-04:00
```

Anonymous, this is mostly fixed as of version 9.2. Please read [http://blog.endpoint.com/2014/04/custom-plans-prepared-statements-in.html](http://blog.endpoint.com/2014/04/custom-plans-prepared-statements-in.html)
